### PR TITLE
sparcv8leon3: add amba pnp support

### PR DESCRIPTION
**Pull request by lukileczo**
JIRA: RTOS-704

## Description

GRLIB components feature Plug and Play information about cores (e.g. address, interrupt number) stored in read-only memory. This PR adds support to read the information about given core through platformctl syscall.

Decided to use MMU bypass to make this feature work seamlessly on both MMU and NOMMU architectures.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: `sparcv8leon3-gr712rc-board`

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: hal/sparcv8leon3/gaisler/ambapp.c [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+#define AMBAPP_APB_OFF ((addr_t)0xff000) /* PnP offset relative to APB bridge base address */
+
+#define AMBAPP_AHB_NMASTERS 16
+#define AMBAPP_AHB_NSLAVES  16
+#define AMBAPP_APB_NSLAVES  16
+
+#define AMBAPP_VEN(id)  ((id) >> 24)
+#define AMBAPP_DEV(id)  (((id) >> 12) & 0xfff)
+#define AMBAPP_VER(id)  (((id) >> 5) & 0x1f)
+#define AMBAPP_IRQN(id) ((id) & 0x1f)
```
[the remark this answers is not in the document]
Comment (lukileczo) [reply hal/sparcv8leon3/gaisler/ambapp.c:36]: no

File: include/arch/sparcv8leon3.h [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+	union {
+		struct {
+			unsigned int *base[4];
+			unsigned int type[4];
+		} ahbInfo;
+		struct {
+			unsigned int *base;
+			unsigned int type;
+		} apbInfo;
+	};
```
Comment (agkaminski) [inline include/arch/sparcv8leon3.h:199]: Anonymous union within struct is a gnu extension (and part of C11). I know we have many examples, but they'll have to go eventually. Perhaps name union `info` and it's elements `ahb` and `apb`?

File: include/arch/sparcv8leon3.h [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+#define CORE_ID_PW2APB    0x03cu
+#define CORE_ID_AHB2PP    0x039u
+#define CORE_ID_GRRM      0x09au
+
+#define AMBA_TYPE_APBIO  0x1u
+#define AMBA_TYPE_AHBMEM 0x2u
+#define AMBA_TYPE_AHBIO  0x3u
+
+
+typedef struct _ambapp_dev_t {
```
[the remark this answers is not in the document]
Comment (lukileczo) [reply include/arch/sparcv8leon3.h:183]: Regarding anonimity - I use this struct in `include/arch/gr716.h` and `gr712rc.h` and I can't include this header there (I use forward declaration in this files), I'd prefer it to stay this way.

[the remark this answers is not in the document]
Comment (agkaminski) [reply include/arch/sparcv8leon3.h:183]: Why you can't do that? It's already been done, e.g. https://github.com/phoenix-rtos/phoenix-rtos-kernel/blob/91fc6c141ae21432031883ef4f39c14c12165c15/hal/armv7a/imx6ull/config.h#L27

[the remark this answers is not in the document]
Comment (agkaminski) [reply include/arch/sparcv8leon3.h:183]: Ok, I understand, yeah, not possible now.

**Review by agkaminski — APPROVED** [review]
LGTM

**Pull request merged**